### ISSUE TITLE
🐛 Bug Report: Website and other other uploads hang while an upload is finalizing

### ISSUE THREAD

**Issue by jwdoll** [stonith404/pingvin-share#660]
### 👟 Reproduction steps

To trigger in one session, upload two files. Say 1 500MB zip and 1 5gb zip. When the 500MB zip is finished uploading and "finalizing", the other download errors, and resumes after the first is finished finalizing. Same result across a variety of chunks sizes.

### 👍 Expected behavior

Finalizing an upload should not block the web client nor interfere with ongoing uploads (I presume it blocks and overflows a buffer). Not a web programmer, but seems like something isn't async that should be. 

### 👎 Actual Behavior

While an upload is finalizing, the site hangs for other clients, and when it finishes, you get an internal server error. It does not appear to interrupt ongoing downloads, fortunately.

**Comment by stonith404** [comment]
Are there any errors in the logs on the server?

**Comment by jwdoll** [comment]
Here's what the frontend looks like as it happens. I'll try to find the backend logs today.

<img width="1275" alt="image" src="https://github.com/user-attachments/assets/95cc392f-2e59-4125-84f7-2aaf052a033f">

**Comment by jwdoll** [comment]
Hmm, sorry, but where's the backend log file? Not seeing it in the file system nor in the db.

**Comment by jwdoll** [comment]
The only thing I see in the docker logs are premature closes, which appear to be from clients navigating away without finishing downloads.

**Comment by jwdoll** [comment]
Yes, I'm sure. Although if you were uploading a single file to two separate shares, either from two tabs or two computers, the second upload to finish would error while the first finalizes.

I didn't capture the 400 error that happens after the 403 in the first screenshot. They seem to always come in pairs.

Here is the 403:
<img width="1075" alt="image" src="https://github.com/user-attachments/assets/5d064ce1-48f1-44a4-9ea0-b3670a1399cc">

And the 400:
<img width="1088" alt="image" src="https://github.com/user-attachments/assets/fc421ebb-95a1-4e8d-85d7-4578aa19446a">

**Comment by stonith404** [comment]
Just to confirm, am I understanding correctly: if you upload the same file to two shares simultaneously, the second upload fails once the first one completes?

Because I'm a bit confused from your reproduction steps:
> To trigger in one session, upload two files. Say 1 500MB zip and 1 5gb zip. When the 500MB zip is finished uploading and "finalizing", the other download errors, and resumes after the first is finished finalizing. Same result across a variety of chunks sizes.

If I'm wrong please explain the reproduction steps step-by-step or create a screen recording.

**Comment by jwdoll** [comment]
It doesn't need to be the same file. The point about two shares is that whenever the server is finalizing a download of sufficient size from any source, any other uploads block/error.

There may be some more relevant errors in the logs than I realized. Seems like deadlock on the sqlite db called through Prisma, though I don't really understand Node's threading model, or lack thereof. I will try to set up an environment and step through. 

Reproduction: Upload 1 1GB file and 1 5GB file. The 1GB file finishes first, and while finalizing produces an error in the second upload/hangs the site.

**Comment by jwdoll** [comment]
Small update - The error isn't presenting with my mac as the host with the backend running in the VS Code debugger. It also doesn't present when I upload to my production host (docker, Ubuntu, HDDs) over wifi. That's leading me to think there is a synchronous IO operation happening when the first file upload completes that blocks long enough on a slow HDD for a fast network connection to overflow a buffer with the second file. Renaming from tmp_chunk causing a physical move? Calling .size requires a re-count? I'll keep experimenting and trying to repro on a different host (especially in case it has something to do with ZFS). 

Thanks for reading. Please let me know if you have any guidance as I'm highly unfamiliar with node.js and the rest of the stack.

**Comment by stonith404** [comment]
Thank you for sharing this detailed information. I just tested the upload on my Mac using an external HDD (formatted as APFS, not ZFS), and I didn’t encounter the error. Are you able to mount another HDD or SSD to your server? If so, you could try setting the Pingvin Share data folder to the mounted drive. This way, we could test if the hard drive might be causing the issue.

You also mentioned an error related to an SQLite deadlock. Could you share more details about that error?

I don’t have much experience with disks, so I’m not sure if a specific operation could cause issues on slower HDDs. But if we narrow down the problem and find that it only happens on slow HDDs, I’ll read into this topic.

**Comment by stonith404** [comment]
Apologies for the delayed response, your message got lost somehow.

I don’t have much experience with different file systems, so I’m not sure I can provide a solid answer about the issue.

That said, your theory seems plausible. If `createZip` somehow blocks writes to the file system, it causes Prisma to time out. Could you test this behavior with two separate ZFS partitions? For example, you could place the database on one partition and the shares folder on the other. If this resolves the issue, it would suggest that ZFS blocks reads while `createZip` is running.

Like I said, I’m not very familiar with file systems, so this suggestion might not make sense. Let me know what you think.

As for the secure cookies issue, this should be fixed in the latest version.

**Comment by stonith404** [comment]
I'll close this because there was no activity. Feel free to reopen this again when you're still experiencing issues.
